Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc

```diff
@@ -0,0 +1,481 @@
+
+  if (verbosity_)
+  {
+    ssLog << "simu: xi = " << xi << ", th_x_phys = " << th_x_phys << ", th_y_phys = " << th_y_phys
+      << ", vtx_lhc_eff_x = " << vtx_lhc_eff_x << ", vtx_lhc_eff_y = " << vtx_lhc_eff_y << std::endl;
+  }
+
+  // check empirical aperture
+  if (useEmpiricalApertures_)
+  {
+    const double xi_th = empiricalAperture_xi0 + th_x_phys * empiricalAperture_a;
+    if (xi > xi_th)
+    {
+      if (verbosity_)
+      {
+        ssLog << "stop because of empirical appertures";
+        edm::LogInfo("CTPPSDirectProtonSimulation") << ssLog.str();
+      }
+
+      return;
+    }
+  }
+
+  // transport the proton into each pot/scoring plane
+  for (const auto &ofp : opticalFunctions)
+  {
+    CTPPSDetId rpId(ofp.first);
+    const unsigned int rpDecId = rpId.arm()*100 + rpId.station()*10 + rpId.rp();
+
+    // first check the arm
+    if (rpId.arm() != arm)
+      continue;
+
+    if (verbosity_)
+      ssLog << "  RP " << rpDecId << std::endl;
+
+    // transport proton
+    LHCInterpolatedOpticalFunctionsSet::Kinematics k_in = { vtx_lhc_eff_x*1E-1, th_x_phys, vtx_lhc_eff_y*1E-1, th_y_phys, xi }; // conversions: mm -> cm
+    LHCInterpolatedOpticalFunctionsSet::Kinematics k_out;
+    ofp.second.transport(k_in, k_out, true);
+
+    double b_x = k_out.x * 1E1, b_y = k_out.y * 1E1; // conversions: cm -> mm
+    double a_x = k_out.th_x, a_y = k_out.th_y;
+
+    // if needed, subtract beam position and angle
+    if (produceHitsRelativeToBeam_)
+    {
+      // determine beam position
+      LHCInterpolatedOpticalFunctionsSet::Kinematics k_be_in = { 0., 0., 0., 0., 0. };
+      LHCInterpolatedOpticalFunctionsSet::Kinematics k_be_out;
+      ofp.second.transport(k_be_in, k_be_out, true);
+
+      a_x -= k_be_out.th_x; a_y -= k_be_out.th_y;
+      b_x -= k_be_out.x * 1E1; b_y -= k_be_out.y * 1E1; // conversions: cm -> mm
+    }
+
+    const double z_scoringPlane = ofp.second.getScoringPlaneZ() * 1E1;  // conversion: cm --> mm
+
+    if (verbosity_)
+    {
+      ssLog << "    proton transported: a_x = " << a_x << " rad, a_y = " << a_y << " rad, b_x = " << b_x <<
+        " mm, b_y = " << b_y << " mm, z = " << z_scoringPlane << " mm" << std::endl;
+    }
+
+    // save scoring plane hit
+    if (produceScoringPlaneHits_)
+        out_tracks.emplace_back(rpId.rawId(), b_x, 0., b_y, 0., 0., 0., 0., 0., 0., CTPPSpixelLocalTrackReconstructionInfo::invalid, 0, 0., 0.);
+
+    // stop if rec hits are not to be produced
+    if (!produceRecHits_)
+      continue;
+
+    // loop over all sensors in the RP
+    for (const auto& detIdInt : geometry.getSensorsInRP(rpId))
+    {
+      CTPPSDetId detId(detIdInt);
+
+      // determine the track impact point (in global coordinates)
+      // !! this assumes that local axes (1, 0, 0) and (0, 1, 0) describe the sensor surface
+      CLHEP::Hep3Vector gl_o = geometry.localToGlobal(detId, CLHEP::Hep3Vector(0, 0, 0));
```

Comment:
The use of CLHEP here is dictated by the geometry class, moving away from CLHEP is not possible unless the geometry is modified as well

Reply:
Thanks @fabiocos! I took note on our TODO list: https://github.com/CTPPS/cmssw/issues/40.